src/cls/rbd/cls_rbd.cc

```diff
@@ -1678,7 +1678,8 @@ int snapshot_add(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
 	        (unsigned long long)snap_id.val);
 	return -EIO;
       }
-      if (snap_meta.name == old_meta.name || snap_meta.id == old_meta.id) {
+      if ((snap_meta.name == old_meta.name || snap_meta.id == old_meta.id)
```

Comment:
The snap id should be unique regardless of the namespace -- so the check should be:

```
if (snap_meta.id == old_meta.id ||
    (snap_meta.name == old_meta.name &&
     snap_meta.snapshot_namespace == old_meta.snapshot_namespace)) {
```